### DISCUSSION TITLE
Monorepo package publish with workspace: scheme

### DISCUSSION THREAD

**Discussion by jxh150535011** [pnpm/pnpm#5418] [Q&A]
我会代码依赖进行如下配置
"xxxx": "workspace:^1.0.0 || ^1.2.3 || ^1.2.3-beta"
"yyyy": "workspace:^1.0.0 || ^1.2.3 || ^1.2.3-beta"

1 本地开发过程中，上述代码通过pnpm 可以正常运行
2 线上构建过程会采用 yarn 方式，也能正常
3 但是这个仓库给其他人采用pnpm 构建的时候却出了异常，因为有时候他可能不用关注 xxxx的包，所以他的本地没有xxxx工程，因此会报workspace项目找不到
所以我想 采用pnpm install的时候，能否判断 本地 workspace 是否存在，如果不存在就使用远程包进行加载 ，跟yarn 一样

**Comment by Jack-Works** [comment]
You need this: https://pnpm.io/workspaces#publishing-workspace-packages to publish your monorepo packages

**Reply by jxh150535011** [reply]
例如 
`"ahooks": "workspace:^1.0.0 || ^3.1.9",`
No matching version found for ahooks@^1.0.0 || ^3.1.9 inside the workspace
不可以工作哦  ，因为ahooks 并不是我本地的代码，而我想让这段代码运行起来，更多的是考虑团队之间协同的便捷。让每个人只需要关注仓库自身相关的依赖就好，不需要把所有的工程拷贝下来

**Reply by Jack-Works** [reply]
https://pnpm.io/workspaces#workspace-protocol-workspace

> By default, pnpm will **link packages from the workspace** if the available packages match the declared ranges. For instance, `foo@1.0.0` is linked into `bar` if `bar` has `"foo": "^1.0.0"` in its dependencies and `foo@1.0.0` is not in the workspace.

In this case, you only need to change the package specifier to `"ahooks": "^3.1.9"`, and it will automatically link to the workspace version if there is one, and if there is none, it will install `ahooks` from the registry.

`workspace:` scheme is designed to _enforce_ the workspace resolution according to the documentation.

**Reply by jxh150535011** [reply]
不知道我是不是记错了，之前发生过这种情况：
采用 "ahooks": "^1.0.0" ，哪天我引用了线上的包后，想再引用本地的包就怎么都无法生效。原因是我的本地version 永远都是1.0.0。因为我的发布是采用线上构建部署方式，本地不需要手动调整版本。

如果按照 "ahooks": "1.0.0 || ^1.0.0" 的方式 优先检查线上模块，如果通过这种方式，我就需要调整本地的version = "1.1.0"，并且依赖项调整为 "ahooks": "1.1.0 || ^1.0.0"  来满足这个模块依赖的方式。这个方式我有时间可以尝试下，但是远没有采用workspace的方式，可以优先引用本地的方式来的直接。

所以我的期望应该有一种方式 能优先检测本地工程，而不是优先检查线上的方式  workspace 就挺好
